src/garage/torch/algos/maml.py

```diff
@@ -0,0 +1,383 @@
+        env (garage.envs.base): A gym environment.
+        policy (garage.torch.policies.base.Policy): Policy.
+        baseline (garage.np.baselines.Baseline): The baseline.
+        meta_optimizer (Union[type, tuple[type, dict]]): Type of optimizer.
+            This can be an optimizer type such as `torch.optim.Adam` or a
+            tuple of type and dictionary, where dictionary contains arguments
+            to initialize the optimizer e.g. `(torch.optim.Adam, {'lr' = 1e-3})`  # noqa: E501
+        meta_batch_size (int): Number of tasks.
+        inner_lr (double): Adaptation learning rate.
+        num_grad_updates (int): Number of adaptation gradient steps.
+        inner_algo (garage.torch.algos.vpg): The inner algorithm used for
+            computing loss.
+
+    """
+
+    def __init__(self,
+                 env,
+                 policy,
+                 baseline,
+                 meta_optimizer=torch.optim.Adam,
+                 meta_batch_size=40,
+                 inner_lr=0.1,
+                 num_grad_updates=1,
+                 inner_algo=None):
+        if inner_algo is None:
+            inner_algo = VPG(env.spec, policy, baseline)
+
+        if policy.vectorized:
+            self.sampler_cls = OnPolicyVectorizedSampler
+        else:
+            self.sampler_cls = BatchSampler
+
+        self.policy = policy
+        self.baseline = baseline
+        self.max_path_length = inner_algo.max_path_length
+        self._env = env
+        self._num_grad_updates = num_grad_updates
+        self._meta_batch_size = meta_batch_size
+        self._inner_algo = inner_algo
+        self._inner_optimizer = DiffSGD(self.policy, lr=inner_lr)
+        self._meta_optimizer = make_optimizer(meta_optimizer,
+                                              policy,
+                                              lr=_Default(1e-3),
+                                              eps=_Default(1e-5))
+        self._episode_reward_mean = [
+            collections.deque(maxlen=100) for _ in range(num_grad_updates + 1)
+        ]
+
+    def train(self, runner):
+        """Obtain samples and start training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            all_samples, all_params = self._obtain_samples(runner)
+            last_return = self.train_once(runner, all_samples, all_params)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, all_samples, all_params):
+        """Train the algorithm once.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[garage.TorchTrajectoryBatch]]): Two
```

Comment:
i think seeing `list[list[BatchDataType]]` is a sign that your batch data type is not very useful. what are you trying to encode here? does it actually need a new type? why is there a list of lists, could it be flat instead?